src/antelope/wallets/ual/MetakeepUAL.ts

```diff
@@ -0,0 +1,534 @@
+    }) {
+        super();
+        this.keys = [publicKey];
+        this.accountName = accountName;
+        this.permission = permission;
+        this.chainId = chainId;
+        this.rpc = rpc;
+        this.accountCreateAPI = accountCreateAPI;
+        this.eosioCore = new APIClient({ url: rpc.endpoint });
+    }
+
+    setReasonCallback(callback: (transaction: any) => string) {
+        this.reasonCallback = callback;
+    }
+
+    handleCatchError(error: any): Error {
+        if (
+            (error as unknown as {status:string}).status === 'USER_REQUEST_DENIED'
+        ) {
+            return new Error('antelope.evm.error_transaction_canceled');
+        } else {
+            return new Error('antelope.evm.error_send_transaction');
+        }
+    }
+
+    /**
+    * @param transaction    The transaction to be signed (a object that matches the RpcAPI structure).
+    */
+    signTransaction = async (originalTransaction: any, options: any = {}): Promise<SignTransactionResponse> => {
+        if (!metakeep) {
+            throw new Error('metakeep is not initialized');
+        }
+
+        try {
+            // expire time in seconds
+            const expireSeconds = 120;
+
+            // Retrieve transaction headers
+            const info = await this.eosioCore.v1.chain.get_info();
+            const header = info.getTransactionHeader(expireSeconds);
+
+            // collect all contract abis
+            const abi_promises = originalTransaction.actions.map((a: { account: NameType; }) =>
+                this.eosioCore.v1.chain.get_abi(a.account),
+            );
+            const responses = await Promise.all(abi_promises);
+            const abis = responses.map(x => x.abi);
+            const abis_and_names = originalTransaction.actions.map((x: { account: any; }, i: number) => ({
+                contract: x.account,
+                abi: abis[i],
+            }));
+
+            // create complete well formed transaction
+            const transaction = Transaction.from(
+                {
+                    ...header,
+                    actions: originalTransaction.actions,
+                },
+                abis_and_names,
+            );
+
+            const transaction_extensions = originalTransaction.transaction_extensions ?? [] as string[];
+            const context_free_actions = originalTransaction.context_free_actions ?? [] as string[];
+            const delay_sec = originalTransaction.delay_sec ?? 0;
+            const max_cpu_usage_ms = originalTransaction.max_cpu_usage_ms ?? 0;
+            const max_net_usage_words = originalTransaction.max_net_usage_words ?? 0;
+            const expiration = originalTransaction.expiration ?? transaction.expiration.toString();
+            const ref_block_num = originalTransaction.ref_block_num ?? transaction.ref_block_num.toNumber();
+            const ref_block_prefix = originalTransaction.ref_block_prefix ?? transaction.ref_block_prefix.toNumber();
```

Comment:
The original transaction may or may not have all the transaction details, containing only the actions. If we are coming from a Greymass Fuel intersection, the original transaction will be complete so we need to use it as it comes in order to sign the same transaction as Fuel did. If not, it's ok to create a brand-new transaction.